File: typescript/src/utils/compute-all.ts
/**
 * Batch compute all applicable ratios from a single financial data object.
 *
 * Usage:
 *   import { computeAll } from 'fin-ratios'
 *
 *   const data = await fetchYahoo('AAPL')
 *   const ratios = computeAll(data)
 *   console.log(ratios.pe)           // 28.3
 *   console.log(ratios.roic)         // 0.55
 *   console.log(ratios.altmanZ)      // { zScore: 4.8, zone: 'safe', ... }
 */

import { safeDivide } from './safe-divide.js'

// Import all ratio functions
import {
  pe, pb, ps, peg, pFcf, evEbitda, evEbit, evRevenue, evFcf, tobinsQ, grahamNumber,
} from '../ratios/valuation/index.js'
import {
  grossMargin, operatingMargin, ebitdaMargin, netProfitMargin,
  roe, roa, nopat, roic, roce, investedCapital,
} from '../ratios/profitability/index.js'
import {
  currentRatio, quickRatio, dso, dio, dpo, cashConversionCycle,
} from '../ratios/liquidity/index.js'
import {
  debtToEquity, netDebtToEquity, netDebtToEbitda,
  debtToAssets, interestCoverageRatio, equityMultiplier,
} from '../ratios/solvency/index.js'
import {
  assetTurnover, receivablesTurnover, inventoryTurnover, payablesTurnover,
} from '../ratios/efficiency/index.js'
import {
  freeCashFlow, fcfYield, fcfMargin, fcfConversion, ocfToSales, capexToRevenue,
} from '../ratios/cashflow/index.js'
import {
  altmanZScore, piotroskiFScore,
} from '../ratios/composite/index.js'

// ── Input type ────────────────────────────────────────────────────────────────

export interface FinancialData {
  // Income Statement
  revenue?: number
  grossProfit?: number
  ebit?: number
  ebitda?: number
  netIncome?: number
  interestExpense?: number
  incomeTaxExpense?: number
  ebt?: number
  cogs?: number
  // Balance Sheet
  totalAssets?: number
  totalEquity?: number
  totalDebt?: number
  currentAssets?: number
  currentLiabilities?: number
  cash?: number
  accountsReceivable?: number
  inventory?: number
  retainedEarnings?: number
  totalLiabilities?: number
  accountsPayable?: number
  longTermDebt?: number
  sharesOutstanding?: number
  // Cash Flow
  operatingCashFlow?: number
  capex?: number
  // Market
  marketCap?: number
  enterpriseValue?: number
  epsGrowthPct?: number
  // Prior period (for Piotroski)
  prior?: Omit<FinancialData, 'prior' | 'marketCap' | 'enterpriseValue' | 'epsGrowthPct'>
}

// ── Output type ───────────────────────────────────────────────────────────────

export interface RatioResults {
  // Valuation
  pe?: number | null
  pb?: number | null
  ps?: number | null
  peg?: number | null
  pFcf?: number | null
  enterpriseValue?: number | null
  evEbitda?: number | null
  evEbit?: number | null
  evRevenue?: number | null
  evFcf?: number | null
  tobinsQ?: number | null
  grahamNumber?: number | null
  // Profitability
  grossMargin?: number | null
  operatingMargin?: number | null
  netMargin?: number | null
  ebitdaMargin?: number | null
  roe?: number | null
  roa?: number | null
  roic?: number | null
  roce?: number | null
  nopat?: number | null
  investedCapital?: number | null
  // Cash Flow
  freeCashFlow?: number | null
  fcfMargin?: number | null
  fcfConversion?: number | null
  ocfToSales?: number | null
  capexToRevenue?: number | null
  fcfYield?: number | null
  // Liquidity
  currentRatio?: number | null
  quickRatio?: number | null
  dso?: number | null
  dio?: number | null
  dpo?: number | null
  cashConversionCycle?: number | null
  // Solvency
  debtToEquity?: number | null
  netDebtToEquity?: number | null
  netDebtToEbitda?: number | null
  debtToAssets?: number | null
  interestCoverage?: number | null
  equityMultiplier?: number | null
  // Efficiency
  assetTurnover?: number | null
  receivablesTurnover?: number | null
  inventoryTurnover?: number | null
  payablesTurnover?: number | null
  // Composite
  altmanZ?: ReturnType<typeof altmanZScore>
  piotroski?: ReturnType<typeof piotroskiFScore> | null
}

// ── Helper ────────────────────────────────────────────────────────────────────

function g(data: FinancialData, key: keyof FinancialData): number {
  const v = data[key]
  if (v === undefined || v === null || typeof v === 'object') return 0
  return (v as number) || 0
}

function gn(data: FinancialData, key: keyof FinancialData): number | undefined {
  const v = data[key]
  if (v === undefined || v === null || typeof v === 'object') return undefined
  return (v as number) || undefined
}

// ── Main function ─────────────────────────────────────────────────────────────

/**
 * Compute all applicable financial ratios from a data object.
 *
 * Returns a flat object with 40+ ratio values. Complex scores (Altman, Piotroski)
 * are nested objects. Returns undefined for unavailable ratios.
 *
 * @param data  Financial data object with income/balance/cashflow/market fields
 * @returns     Flat object of ratio name → value
 *
 * @example
 *   const data = await fetchYahoo('AAPL')
 *   const r = computeAll(data)
 *   console.log(r.grossMargin)  // 0.433
 *   console.log(r.roic)         // 0.55
 *   console.log(r.altmanZ?.zone) // 'safe'
 */
export function computeAll(data: FinancialData): RatioResults {
  const result: RatioResults = {}

  const revenue = g(data, 'revenue')
  const grossProfit = g(data, 'grossProfit')
  const ebit = g(data, 'ebit')
  const ebitda = g(data, 'ebitda')
  const netIncome = g(data, 'netIncome')
  const totalAssets = g(data, 'totalAssets')
  const totalEquity = g(data, 'totalEquity')
  const totalDebt = g(data, 'totalDebt')
  const currentAssets = g(data, 'currentAssets')
  const currentLiabilities = g(data, 'currentLiabilities')
  const cash = g(data, 'cash')
  const accountsReceivable = g(data, 'accountsReceivable')
  const inventory = g(data, 'inventory')
  const marketCap = g(data, 'marketCap')
  const operatingCashFlow = g(data, 'operatingCashFlow')
  const capexRaw = g(data, 'capex')
  const interestExpense = g(data, 'interestExpense')
  const incomeTaxExpense = g(data, 'incomeTaxExpense')
  const ebt = g(data, 'ebt')
  const cogs = g(data, 'cogs')
  const accountsPayable = g(data, 'accountsPayable')
  const retainedEarnings = g(data, 'retainedEarnings')
  const totalLiabilities = g(data, 'totalLiabilities')
  const sharesOutstanding = g(data, 'sharesOutstanding')

  const fcf = freeCashFlow({ operatingCashFlow, capex: capexRaw })
  const ev = data.enterpriseValue || (marketCap + totalDebt - cash) || undefined

  const taxRate = ebt ? Math.min(Math.max(safeDivide(incomeTaxExpense, ebt) ?? 0.21, 0), 0.5) : 0.21
  const nopatVal = nopat({ ebit, taxRate })
  const icVal = investedCapital({ totalEquity, totalDebt, cash })

  // ── Valuation ───────────────────────────────────────────────────────────────
  result.pe = pe({ marketCap, netIncome })
  result.pb = pb({ marketCap, totalEquity })
  result.ps = ps({ marketCap, revenue })
  if (result.pe && data.epsGrowthPct) {
    result.peg = peg({ peRatio: result.pe, epsGrowthRatePercent: data.epsGrowthPct })
  }
  if (operatingCashFlow) result.pFcf = pFcf({ marketCap, operatingCashFlow, capex: capexRaw })
  result.enterpriseValue = ev ?? null
  if (ev && ebitda) result.evEbitda = evEbitda({ enterpriseValue: ev, ebitda })
  if (ev && ebit) result.evEbit = evEbit({ enterpriseValue: ev, ebit })
  if (ev && revenue) result.evRevenue = evRevenue({ enterpriseValue: ev, revenue })
  if (ev && fcf) result.evFcf = evFcf({ enterpriseValue: ev, freeCashFlow: fcf })
  result.tobinsQ = tobinsQ({ marketCap, totalDebt, totalAssets })
  if (sharesOutstanding) {
    const eps = safeDivide(netIncome, sharesOutstanding)
    const bvps = safeDivide(totalEquity, sharesOutstanding)
    if (eps && bvps) result.grahamNumber = grahamNumber({ eps, bookValuePerShare: bvps })
  }

  // ── Profitability ────────────────────────────────────────────────────────────
  result.grossMargin = grossMargin({ grossProfit, revenue })
  result.operatingMargin = operatingMargin({ ebit, revenue })
  result.netMargin = netProfitMargin({ netIncome, revenue })
  if (ebitda) result.ebitdaMargin = ebitdaMargin({ ebitda, revenue })
  result.roe = roe({ netIncome, avgTotalEquity: totalEquity })
  result.roa = roa({ netIncome, avgTotalAssets: totalAssets })
  result.nopat = nopatVal
  result.investedCapital = icVal
  if (nopatVal && icVal) result.roic = roic({ nopat: nopatVal, investedCapital: icVal })
  result.roce = roce({ ebit, totalAssets, currentLiabilities })

  // ── Cash Flow ────────────────────────────────────────────────────────────────
  result.freeCashFlow = fcf ?? null
  if (fcf) {
    result.fcfMargin = fcfMargin({ freeCashFlow: fcf, revenue })
    result.fcfConversion = fcfConversion({ freeCashFlow: fcf, netIncome })
    if (marketCap) result.fcfYield = fcfYield({ freeCashFlow: fcf, marketCap })
  }

  result.ocfToSales = ocfToSales({ operatingCashFlow, revenue })
  result.capexToRevenue = capexToRevenue({ capex: capexRaw, revenue })

  // ── Liquidity ────────────────────────────────────────────────────────────────
  result.currentRatio = currentRatio({ currentAssets, currentLiabilities })
  result.quickRatio = quickRatio({ cash, shortTermInvestments: 0, accountsReceivable, currentLiabilities })
  result.dso = dso({ accountsReceivable, revenue })
  if (cogs) {
    result.dio = dio({ inventory, cogs })
    result.dpo = dpo({ accountsPayable, cogs })
  }
  if (result.dso != null) {
    result.cashConversionCycle = cashConversionCycle({
      dso: result.dso,
      dio: result.dio ?? 0,
      dpo: result.dpo ?? 0,
    })
  }

  // ── Solvency ─────────────────────────────────────────────────────────────────
  result.debtToEquity = debtToEquity({ totalDebt, totalEquity })
  result.netDebtToEquity = netDebtToEquity({ totalDebt, cash, totalEquity })
  if (ebitda) result.netDebtToEbitda = netDebtToEbitda({ totalDebt, cash, ebitda })
  result.debtToAssets = debtToAssets({ totalDebt, totalAssets })
  result.interestCoverage = interestCoverageRatio({ ebit, interestExpense })
  result.equityMultiplier = equityMultiplier({ totalAssets, totalEquity })

  // ── Efficiency ───────────────────────────────────────────────────────────────
  result.assetTurnover = assetTurnover({ revenue, avgTotalAssets: totalAssets })
  result.receivablesTurnover = receivablesTurnover({ revenue, avgAccountsReceivable: accountsReceivable })
  if (cogs) {
    result.inventoryTurnover = inventoryTurnover({ cogs, avgInventory: inventory })
    result.payablesTurnover = payablesTurnover({ cogs, avgAccountsPayable: accountsPayable })
  }

  // ── Composite ────────────────────────────────────────────────────────────────
  if (totalAssets && totalLiabilities) {
    result.altmanZ = altmanZScore({
      workingCapital: currentAssets - currentLiabilities,
      retainedEarnings,
      ebit,
      marketCap,
      totalLiabilities,
      totalAssets,
      revenue,
    })
  }

  if (data.prior) {
    const p = data.prior
    const pGet = (key: keyof typeof p): number => {
      const v = p[key]
      return (typeof v === 'number' ? v : 0) || 0
    }
    try {
      result.piotroski = piotroskiFScore({
        current: {
          netIncome,
          totalAssets,
          operatingCashFlow,
          longTermDebt: g(data, 'longTermDebt'),
          currentAssets,
          currentLiabilities,
          sharesOutstanding,
          grossProfit,
          revenue,
        },
        prior: {
          netIncome: pGet('netIncome'),
          totalAssets: pGet('totalAssets'),
          longTermDebt: pGet('longTermDebt'),
          currentAssets: pGet('currentAssets'),
          currentLiabilities: pGet('currentLiabilities'),
          sharesOutstanding: pGet('sharesOutstanding'),
          grossProfit: pGet('grossProfit'),
          revenue: pGet('revenue'),
        },
      })
    } catch {
      result.piotroski = null
    }
  }

  return result
}

File: python/fin_ratios/utils/compute_all.py
"""
Batch compute all applicable ratios from a single data object.

Usage:
    from fin_ratios.utils.compute_all import compute_all
    from fin_ratios.fetchers.yahoo import fetch_yahoo

    data = fetch_yahoo('AAPL')
    ratios = compute_all(data)
    print(ratios['pe'])           # 28.3
    print(ratios['roic'])         # 0.55
    print(ratios['piotroski'])    # {'score': 8, 'interpretation': ...}
"""

from __future__ import annotations

from typing import Any, Optional

from fin_ratios._utils import safe_divide
import fin_ratios as _fr


def compute_all(data: Any, prior: Optional[Any] = None) -> dict[str, Any]:
    """
    Compute all applicable ratios from a fetched data object.

    Args:
        data:  Any object with financial attributes (YahooFinancialData,
               FmpFinancialData, etc.) or a plain dict.
        prior: Optional prior-period data for ratios that need year-over-year
               comparison (Piotroski, Beneish, growth ratios).

    Returns:
        Flat dict of ratio name → value. Complex ratios (Piotroski, Altman, etc.)
        are nested dicts. Returns None for unavailable or divide-by-zero cases.

    Example:
        data = fetch_yahoo('AAPL')
        r = compute_all(data)
        # r['pe'] = 28.3
        # r['altman_z'] = {'z_score': 4.8, 'zone': 'safe', ...}
        # r['piotroski'] = {'score': 8, ...}
    """
    d = _DataAccessor(data)
    result: dict[str, Any] = {}

    # ── Valuation ─────────────────────────────────────────────────────────────
    result["pe"] = _fr.pe(market_cap=d.market_cap, net_income=d.net_income)
    result["pb"] = _fr.pb(market_cap=d.market_cap, total_equity=d.total_equity)
    result["ps"] = _fr.ps(market_cap=d.market_cap, revenue=d.revenue)
    result["peg"] = (
        _fr.peg(pe_ratio=result["pe"], eps_growth_rate_pct=d.eps_growth_pct)
        if result["pe"] and d.eps_growth_pct
        else None
    )

    fcf = _fr.free_cash_flow(operating_cash_flow=d.operating_cash_flow, capex=d.capex)
    result["free_cash_flow"] = fcf
    result["p_fcf"] = (
        _fr.p_fcf(market_cap=d.market_cap, operating_cash_flow=d.operating_cash_flow, capex=d.capex)
        if d.operating_cash_flow
        else None
    )

    ev = d.enterprise_value or (d.market_cap + d.total_debt - d.cash)
    result["enterprise_value"] = ev
    result["ev_ebitda"] = _fr.ev_ebitda(ev=ev, ebitda=d.ebitda) if d.ebitda else None
    result["ev_ebit"] = _fr.ev_ebit(ev=ev, ebit=d.ebit) if d.ebit else None
    result["ev_revenue"] = _fr.ev_revenue(ev=ev, revenue=d.revenue) if d.revenue else None
    result["ev_fcf"] = _fr.ev_fcf(ev=ev, free_cash_flow=fcf) if fcf else None
    result["tobin_q"] = _fr.tobin_q(
        market_cap=d.market_cap, total_debt=d.total_debt, total_assets=d.total_assets
    )
    result["graham_number"] = (
        _fr.graham_number(
            eps=safe_divide(d.net_income, d.shares_outstanding),
            book_value_per_share=safe_divide(d.total_equity, d.shares_outstanding),
        )
        if d.shares_outstanding
        else None
    )

    # ── Profitability ─────────────────────────────────────────────────────────
    result["gross_margin"] = _fr.gross_margin(gross_profit=d.gross_profit, revenue=d.revenue)
    result["operating_margin"] = _fr.operating_margin(ebit=d.ebit, revenue=d.revenue)
    result["net_margin"] = _fr.net_profit_margin(net_income=d.net_income, revenue=d.revenue)
    result["ebitda_margin"] = (
        _fr.ebitda_margin(ebitda=d.ebitda, revenue=d.revenue) if d.ebitda else None
    )
    result["roe"] = _fr.roe(net_income=d.net_income, avg_total_equity=d.total_equity)
    result["roa"] = _fr.roa(net_income=d.net_income, avg_total_assets=d.total_assets)
    result["roce"] = _fr.roce(
        ebit=d.ebit, total_assets=d.total_assets, current_liabilities=d.current_liabilities
    )

    tax_rate = safe_divide(d.income_tax_expense, d.ebt) if d.ebt else 0.21
    nopat_val = _fr.nopat(ebit=d.ebit, tax_rate=min(max(tax_rate or 0, 0), 0.5))
    ic_val = _fr.invested_capital(total_equity=d.total_equity, total_debt=d.total_debt, cash=d.cash)
    result["nopat"] = nopat_val
    result["invested_capital"] = ic_val
    result["roic"] = _fr.roic(nopat_value=nopat_val, invested_capital=ic_val)

    # ── Cash Flow ─────────────────────────────────────────────────────────────
    result["fcf_margin"] = _fr.fcf_margin(fcf=fcf, revenue=d.revenue) if fcf else None
    result["fcf_conversion"] = _fr.fcf_conversion(fcf=fcf, net_income=d.net_income) if fcf else None
    result["ocf_to_sales"] = _fr.ocf_to_sales(
        operating_cash_flow=d.operating_cash_flow, revenue=d.revenue
    )
    result["capex_to_revenue"] = _fr.capex_to_revenue(capex=d.capex, revenue=d.revenue)
    result["fcf_yield"] = _fr.fcf_yield(fcf=fcf, market_cap=d.market_cap) if fcf else None

    # ── Liquidity ─────────────────────────────────────────────────────────────
    result["current_ratio"] = _fr.current_ratio(
        current_assets=d.current_assets, current_liabilities=d.current_liabilities
    )
    result["quick_ratio"] = _fr.quick_ratio(
        cash=d.cash,
        short_term_investments=0,
        accounts_receivable=d.accounts_receivable,
        current_liabilities=d.current_liabilities,
    )
    result["dso"] = _fr.dso(accounts_receivable=d.accounts_receivable, revenue=d.revenue)
    result["dio"] = _fr.dio(inventory=d.inventory, cogs=d.cogs) if d.cogs else None
    result["dpo"] = _fr.dpo(accounts_payable=d.accounts_payable, cogs=d.cogs) if d.cogs else None

    dso_v, dio_v, dpo_v = result["dso"], result["dio"], result["dpo"]
    result["cash_conversion_cycle"] = (
        _fr.cash_conversion_cycle(dso_days=dso_v, dio_days=dio_v or 0, dpo_days=dpo_v or 0)
        if dso_v is not None
        else None
    )

    # ── Solvency ──────────────────────────────────────────────────────────────
    result["debt_to_equity"] = _fr.debt_to_equity(
        total_debt=d.total_debt, total_equity=d.total_equity
    )
    result["net_debt_to_equity"] = _fr.net_debt_to_equity(
        total_debt=d.total_debt, cash=d.cash, total_equity=d.total_equity
    )
    result["net_debt_to_ebitda"] = (
        _fr.net_debt_to_ebitda(total_debt=d.total_debt, cash=d.cash, ebitda=d.ebitda)
        if d.ebitda
        else None
    )
    result["debt_to_assets"] = _fr.debt_to_assets(
        total_debt=d.total_debt, total_assets=d.total_assets
    )
    result["interest_coverage"] = _fr.interest_coverage_ratio(
        ebit=d.ebit, interest_expense=d.interest_expense
    )
    result["equity_multiplier"] = _fr.equity_multiplier(
        total_assets=d.total_assets, total_equity=d.total_equity
    )

    # ── Efficiency ────────────────────────────────────────────────────────────
    result["asset_turnover"] = _fr.asset_turnover(
        revenue=d.revenue, avg_total_assets=d.total_assets
    )
    result["receivables_turnover"] = _fr.receivables_turnover(
        revenue=d.revenue, avg_accounts_receivable=d.accounts_receivable
    )
    result["inventory_turnover"] = (
        _fr.inventory_turnover(cogs=d.cogs, avg_inventory=d.inventory) if d.cogs else None
    )
    result["payables_turnover"] = (
        _fr.payables_turnover(cogs=d.cogs, avg_accounts_payable=d.accounts_payable)
        if d.cogs
        else None
    )

    # ── Composite Scores ──────────────────────────────────────────────────────
    try:
        altman = _fr.altman_z_score(
            working_capital=d.current_assets - d.current_liabilities,
            retained_earnings=d.retained_earnings,
            ebit=d.ebit,
            market_cap=d.market_cap,
            total_liabilities=d.total_liabilities,
            total_assets=d.total_assets,
            revenue=d.revenue,
        )
        result["altman_z"] = altman
    except Exception:
        result["altman_z"] = None

    # Prior-period ratios (Piotroski, Beneish, growth)
    if prior is not None:
        p = _DataAccessor(prior)
        try:
            result["piotroski"] = _fr.piotroski_f_score(
                current_net_income=d.net_income,
                current_total_assets=d.total_assets,
                current_operating_cf=d.operating_cash_flow,
                current_long_term_debt=d.long_term_debt,
                current_current_assets=d.current_assets,
                current_current_liabilities=d.current_liabilities,
                current_shares_outstanding=d.shares_outstanding,
                current_gross_profit=d.gross_profit,
                current_revenue=d.revenue,
                prior_net_income=p.net_income,
                prior_total_assets=p.total_assets,
                prior_long_term_debt=p.long_term_debt,
                prior_current_assets=p.current_assets,
                prior_current_liabilities=p.current_liabilities,
                prior_shares_outstanding=p.shares_outstanding,
                prior_gross_profit=p.gross_profit,
                prior_revenue=p.revenue,
            )
        except Exception:
            result["piotroski"] = None

        try:
            from fin_ratios.ratios.composite import montier_c_score

            result["montier_c"] = montier_c_score(
                current_net_income=d.net_income,
                current_operating_cash_flow=d.operating_cash_flow,
                current_accounts_receivable=d.accounts_receivable,
                current_revenue=d.revenue,
                current_inventory=d.inventory,
                current_cogs=d.cogs or d.revenue - d.gross_profit,
                current_cash=d.cash,
                current_total_assets=d.total_assets,
                current_long_term_debt=d.long_term_debt,
                current_gross_profit=d.gross_profit,
                prior_accounts_receivable=p.accounts_receivable,
                prior_revenue=p.revenue,
                prior_inventory=p.inventory,
                prior_cogs=p.cogs or p.revenue - p.gross_profit,
                prior_cash=p.cash,
                prior_total_assets=p.total_assets,
                prior_long_term_debt=p.long_term_debt,
                prior_gross_profit=p.gross_profit,
            )
        except Exception:
            result["montier_c"] = None

        try:
            result["revenue_growth"] = _fr.revenue_growth(
                current_revenue=d.revenue, prior_revenue=p.revenue
            )
        except Exception:
            result["revenue_growth"] = None

    # ── Health Score ──────────────────────────────────────────────────────────
    try:
        from fin_ratios.utils.health_score import health_score

        piotroski_score_val = result.get("piotroski", {})
        piotroski_int = (
            piotroski_score_val.get("score") if isinstance(piotroski_score_val, dict) else None
        )
        altman_z_val = result.get("altman_z", {})
        altman_z_float = altman_z_val.get("z_score") if isinstance(altman_z_val, dict) else None
        result["health_score"] = health_score(
            piotroski_score=piotroski_int,
            altman_z=altman_z_float,
            roic=result.get("roic"),
            net_margin=result.get("net_margin"),
            gross_margin=result.get("gross_margin"),
            fcf_conversion=result.get("fcf_conversion"),
            fcf_margin=result.get("fcf_margin"),
            debt_to_equity=result.get("debt_to_equity"),
            current_ratio=result.get("current_ratio"),
            interest_coverage=result.get("interest_coverage"),
        )
    except Exception:
        result["health_score"] = None

    return result


class _DataAccessor:
    """Uniform attribute access for any data object or dict."""

    def __init__(self, data: Any):
        self._data = data

    def __getattr__(self, name: str) -> float:
        d = object.__getattribute__(self, "_data")
        if isinstance(d, dict):
            return d.get(name, 0.0) or 0.0
        return getattr(d, name, 0.0) or 0.0
